x-pack/plugins/security_solution/public/common/components/exceptions/use_fetch_or_create_rule_exception_list.tsx

```diff
@@ -0,0 +1,165 @@
+      const exceptionListToCreate: CreateExceptionListSchema = {
+        name: ruleResponse.name,
+        description: ruleResponse.description,
+        type: exceptionListType,
+        namespace_type: exceptionListType === 'endpoint' ? 'agnostic' : 'single',
+        _tags: undefined,
+        tags: undefined,
+        list_id: exceptionListType === 'endpoint' ? 'endpoint_list' : undefined,
+        meta: undefined,
+      };
+      try {
+        const newExceptionList = await addExceptionList({
+          http,
+          list: exceptionListToCreate,
+          signal: abortCtrl.signal,
+        });
+        return Promise.resolve(newExceptionList);
+      } catch (error) {
+        return Promise.reject(error);
+      }
+    }
+    async function createAndAssociateExceptionList(
+      ruleResponse: Rule
+    ): Promise<ExceptionListSchema> {
+      const newExceptionList = await createExceptionList(ruleResponse);
+
+      const newExceptionListReference = {
+        id: newExceptionList.id,
+        type: newExceptionList.type,
+        namespace_type: newExceptionList.namespace_type,
+      };
+      const newExceptionListReferences: ListArray = [
+        ...(ruleResponse.exceptions_list ?? []),
+        newExceptionListReference,
+      ];
+
+      await patchRule({
+        ruleProperties: {
+          rule_id: ruleResponse.rule_id,
+          exceptions_list: newExceptionListReferences,
+        },
+        signal: abortCtrl.signal,
+      });
+
+      return Promise.resolve(newExceptionList);
+    }
+
+    async function fetchRule(): Promise<Rule> {
+      return fetchRuleById({
+        id: ruleId,
+        signal: abortCtrl.signal,
+      });
+    }
+
+    async function fetchRuleExceptionLists(ruleResponse: Rule): Promise<ExceptionListSchema[]> {
+      const exceptionListReferences = ruleResponse.exceptions_list as ListArray;
+      if (exceptionListReferences && exceptionListReferences.length > 0) {
+        const exceptionListPromises = exceptionListReferences.map(
+          (exceptionListReference: List) => {
+            return fetchExceptionListById({
+              http,
+              id: exceptionListReference.id,
+              namespaceType: exceptionListReference.namespace_type,
+              signal: abortCtrl.signal,
+            });
+          }
+        );
+        return Promise.all(exceptionListPromises);
+      } else {
+        return Promise.resolve([]);
+      }
+    }
+
+    async function fetchOrCreateRuleExceptionList() {
+      try {
+        setIsLoading(true);
+        const ruleResponse = await fetchRule();
+        const exceptionLists = await fetchRuleExceptionLists(ruleResponse);
+
+        let exceptionListToUse: ExceptionListSchema;
+        const matchingList = exceptionLists.find((list) => {
+          if (exceptionListType === 'endpoint') {
+            return list.type === exceptionListType && list.list_id === 'endpoint_list';
```

Comment:
I might check with @FrankHassanabad , but almost certain that we shouldn't create a `list_id` in the client. We should let the server create it and then reference it here using the response from the creation.

Reply:
We need to think about the right behavior is here when there is no exception list present. We can iterate on it.